### ISSUE TITLE
Remove `GemmaTransformerDecoder`, and use `TiedEmbeddingTransformerDecoder` in all Gemma models.

### ISSUE THREAD

**Issue by SalmanMohammadi** [pytorch/torchtune#1447]
Thanks to one of our contributors, as part of the [Qwen2](https://github.com/pytorch/torchtune/tree/main/torchtune/models/qwen2) torchtune implementation we now support a `TiedEmbeddingTransformerDecoder` which can be generally used for models which have tied embedding-output weights. This means we no longer need a Gemma-specific decoder class.

To do this, I would:
1) For `gemma` and `lora_gemma` [component builders](https://github.com/pytorch/torchtune/blob/main/torchtune/models/gemma/_component_builders.py), return an instance of `TiedEmbeddingTransformerDecoder` instead of `GemmaTransformerDecoder`. See the Qwen2 [component builders](https://github.com/pytorch/torchtune/blob/4fba6cd44c8f24fe4ea3a2cf3a9c03e12fd9f847/torchtune/models/qwen2/_component_builders.py#L260) for reference.
2) Remove the `torchtune.models.gemma.transformer.py` file, and ensure all references to` GemmaTransformerDecoder` are either removed (for example, in the `__init__.py`), or changed to `TiedEmbeddingTransformerDecoder`. This should also mean updating all typing in the `_model_builders.py` and `_component_builders.py` to correctly return `TiedEmbeddingTransformerDecoder` instead of `GemmaTransformerDecoder
3) Ensure all tests pass, and verify the implementation is correct by running a recipe with at least one base Gemma model, and one LoRA gemma model, ensuring things look sensible (i.e. generations are working correctly, loss curves go down, etc.).

Should address #1446

**Comment by ebsmothers** [comment]
Also just going to throw this out there: @pbontrager had suggested that we can do away with `TiedEmbeddingTransformerDecoder` entirely and instead do something like (crappy diff snippet from our Llama2 builder as an example)

```
    tok_embeddings = nn.Embedding(vocab_size, embed_dim)
    # output_proj = nn.Linear(embed_dim, vocab_size, bias=False)
    output_proj = lambda x: F.linear(x, tok_embeddings.weight)
    return TransformerDecoder(
        tok_embeddings=tok_embeddings,
        layers=layer,
        num_layers=num_layers,
        max_seq_len=max_seq_len,
        num_heads=num_heads,
        head_dim=head_dim,
        norm=RMSNorm(embed_dim, eps=norm_eps),
        output=output_proj,
    )
```

Possibly without a lambda and maybe with a proper function. Main open question is whether this works with FSDP and checkpointing

**Comment by SalmanMohammadi** [comment]
I personally like this, and it doesn't mess with checkpointing when I last tried it out - not sure about FSDP. Happy to take this down and put a new issue up if we agree on this - it'd save a lot of testing and maintenance for what is essentially just a copy of `TransformerDecoder`.

**Comment by ebsmothers** [comment]
I think we definitely agree that we should try it out. And if we find that everything works nicely with this approach it'd be much cleaner (and I know Philip agrees 😅)

**Comment by SalmanMohammadi** [comment]
Worth just adding in the necessary cache setup fn to the Gemma decoder now to unblock #1446? It currently seems like generation is broken for our Gemma models.
